### PR TITLE
stages/files: do not fail if links are correct

### PR DESCRIPTION
**Pull request by ajeddeloh**
If a link already exists that matches the specified link, do not fail.

WIP needs tests, docs

### CODE REVIEW

**Comment by ajeddeloh** [comment]
Depends on https://github.com/coreos/ignition/pull/794

File: internal/exec/stages/files/filesystemEntries.go [hunk 1]
```diff
@@ -140,6 +142,42 @@ func (tmp linkEntry) node() types.Node {
...
+			if target, err := os.Readlink(s.Path); err != nil {
+				return fmt.Errorf("error reading link at %s: %v", s.Path, err)
+			} else if filepath.Clean(target) != s.Target {
+				return fmt.Errorf("error creating symlink %s: a symlink exists at that path but points to %s, not %s and overwrite is false", s.Path, target, s.Target)
+			} else {
+				l.Info("Symlink %s to %s already exists, doing nothing", s.Path, s.Target)
+				return nil
+			}
+		}
+		return fmt.Errorf("error creating symlink %s: a non-symlink already exists that that path and overwrite is false", s.Path)
```
Comment (jlebon) [inline internal/exec/stages/files/filesystemEntries.go]: that that

File: internal/exec/stages/files/filesystemEntries.go [hunk 2]
```diff
@@ -140,6 +142,42 @@ func (tmp linkEntry) node() types.Node {
...
+			return fmt.Errorf("error creating hard link %s: a file already exists at that path but is not the target and overwrite is false", s.Path)
+		}
+		l.Info("Hardlink %s to %s already exists, doing nothing", s.Path, s.Target)
+		return nil
+	case !hard:
+		// if the existing file is a symlink, check that its target is correct
+		if st.Mode()&os.ModeSymlink != 0 {
+			if target, err := os.Readlink(s.Path); err != nil {
+				return fmt.Errorf("error reading link at %s: %v", s.Path, err)
+			} else if filepath.Clean(target) != s.Target {
```
Comment (jlebon) [inline internal/exec/stages/files/filesystemEntries.go]: Sanity-check: is `s.Target` also already cleaned? Or should this be `filepath.Clean(target) != filepath.Clean(s.Target)`?

File: doc/configuration-v3_0.md [hunk 3]
```diff
@@ -81,7 +81,7 @@ The Ignition configuration is a JSON document conforming to the following specif
       * **_name_** (string): the group name of the owner.
   * **_links_** (list of objects): the list of links to be created. Every file, directory, and link must have a unique `path`.
     * **path** (string): the absolute path to the link
-    * **_overwrite_** (boolean): whether to delete preexisting nodes at the path. Defaults to false.
+    * **_overwrite_** (boolean): whether to delete preexisting nodes at the path. Defaults to false. If overwrite is false and a matching link exists at the path, Ignition will only set the owner and group.
```
Comment (jlebon) [inline doc/configuration-v3_0.md]: Minor: in `files` and `directories`, we finish the blurb with `Defaults to false`. Should probably be consistent here too.

File: internal/exec/stages/files/filesystemEntries.go [hunk 4]
```diff
@@ -45,6 +45,8 @@ func (s *stage) createFilesystemsEntries(config types.Config) error {
 
 // filesystemEntry represent a thing that knows how to create itself.
 type filesystemEntry interface {
+	// create creates the entry if specified. It assumes that if overwrite=true then any existing
```
[the remark this answers is not in the document]
Comment (ajeddeloh) [reply internal/exec/stages/files/filesystemEntries.go]: That's the golang comment style. `<functionname> <verbs> <stuff>`

File: internal/exec/stages/files/filesystemEntries.go (hunk 2 above)
Comment (ajeddeloh) [reply internal/exec/stages/files/filesystemEntries.go]: Thought it was, turns out it wasn't. Adding it here (and will PR with it to require clean paths for target).

**Comment by ajeddeloh** [comment]
Updated with fixes, except the go comment thing because that's the way golang wants it.

File: doc/configuration-v3_0.md [hunk 5]
```diff
@@ -81,7 +81,7 @@ The Ignition configuration is a JSON document conforming to the following specif
       * **_name_** (string): the group name of the owner.
   * **_links_** (list of objects): the list of links to be created. Every file, directory, and link must have a unique `path`.
     * **path** (string): the absolute path to the link
-    * **_overwrite_** (boolean): whether to delete preexisting nodes at the path. Defaults to false.
+    * **_overwrite_** (boolean): whether to delete preexisting nodes at the path. If overwrite is false and a matching link exists at the path, Ignition will only set the owner and group. Defaults to false
```
Comment (jlebon) [inline doc/configuration-v3_0.md]: Minor: missing final period here.

**Pull request merged**